src/database/sql/sql.go

```diff
@@ -1048,6 +1048,49 @@ func (db *DB) SetConnMaxIdleTime(d time.Duration) {
 	db.startCleanerLocked()
 }
 
+// SetMaxBadConnRetries sets the number of maximum retries if the driver returns
+// driver.ErrBadConn to signal a broken connection before forcing a new
+// connection to be opened.
+//
+// If n <= 0, no retry when driver.ErrBadConn happen
+//
+// The default max bad retries connections is currently 2. This may change in
+// a future release.
+func (db *DB) SetMaxBadConnRetries(n int) {
+	db.mu.Lock()
+	db.maxBadConnRetries = n
+	db.mu.Unlock()
+}
+
+const defaultMaxBadConnRetries = 2
+
+func (db *DB) retry(fn func(strategy connReuseStrategy) error) error {
+	var err error
+	var isBadConn = true
+
+	maxBadConnRetries := db.maxBadConnRetries
+	if maxBadConnRetries == 0 {
+		maxBadConnRetries = defaultMaxBadConnRetries
+	}
+
+	for i := 0; i < maxBadConnRetries; i++ {
+		err = fn(cachedOrNewConn)
+		isBadConn = err != nil && errors.Is(err, driver.ErrBadConn)
+		if !isBadConn {
+			break
```

Comment:
I suggest to return `err` here and `nil` at the end of the function.

Reply:
refactored the retry method, thank you for your review.